Sources/NIO/Codec.swift

```diff
@@ -96,59 +96,83 @@ private extension ChannelHandlerContext {
 
 extension ByteToMessageDecoder {
 
+    /// Decode in a loop until there is nothing more to decode.
+    private func decodeLoop(ctx: ChannelHandlerContext, decodeFunc: (ChannelHandlerContext, inout ByteBuffer) throws -> DecodingState) {
+        assert(self.cumulationBuffer != nil)
+        ctx.withThrowingToFireErrorAndClose {
+            while true {
+                var buffer = self.cumulationBuffer!
+                let readable = buffer.readableBytes
+                let writerIndex = buffer.writerIndex
+                let result = try decodeFunc(ctx, &buffer)
+                if self.cumulationBuffer != nil {
+                    self.cumulationBuffer!.moveReaderIndex(forwardBy: readable - buffer.readableBytes)
```

Comment:
Is this right? We’re passing a local copy of the buffer, but then mutating the cumulation buffer. Why not just preserve the buffer returned from the call to decode, especially as we’re bothering to pass it `inout`.

Reply:
I think it’s right but I am all ears for better ideas. Can you show me some code with your suggestion as I don’t understand what exactly you suggest here

Reply:
Basically I'm just saying why not just do `self.cumulationBuffer = buffer`?

Reply:
Ah, I know why: the `cumulationBuffer` may have been mutated. That's awkward.

Reply:
Hrm, I'm inclined to say that we should probably pass `&buffer` as a slice instead, and then come up with some custom logic to reconcile the changes from that slice with the parent.